### PR TITLE
Pcm capibility clean up

### PR DESCRIPTION
**Pull request by keyonjie**
refine the supported formats of the PCM capability, it is up to the pipeline format:
    
    for s16_le pipeline, only S16_LE is supported.
    for s24_le pipeline, both S16_LE and S24_LE are supported.
    for s32_le pipeline, all S16_LE, S24_LE and S32_LE are supported.

### CODE REVIEW

File: tools/topology/sof/pipe-amp-ref-capture.m4 [hunk 1]
```diff
@@ -106,5 +106,5 @@ indir(`define', concat(`PIPELINE_PCM_', PIPELINE_ID), Echo Reference Capture PCM
 
 
 # PCM capabilities supported by FW
-PCM_CAPABILITIES(Echo Reference Capture PCM_ID, `S32_LE,S24_LE,S16_LE', 48000, 48000, 2, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
+PCM_CAPABILITIES(Echo Reference Capture PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT), 48000, 48000, 2, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
```
Comment (ranj063) [inline tools/topology/sof/pipe-amp-ref-capture.m4:109]: @keyonjie for my education, what is the pcm_converter?

Comment (ranj063) [reply tools/topology/sof/pipe-amp-ref-capture.m4:109]: And what problem does this help fix?

Comment (keyonjie) [reply tools/topology/sof/pipe-amp-ref-capture.m4:109]: @ranj063 pcm_converter is a converter help to convert format to the same one of the dai format added inside the firmware. With it, the volume/pga component is not mandatory when the dai component is configured to different format with others (e.g. host).

This PR will help fix issue where we are claiming the wrong supported formats in the PCM capability, e.g. our host buffer is allocated as S16_LE so we can support S16_LE only, but with today's code it is `S32_LE,S24_LE,S16_LE' in the pcm capability, so end user can run it with S24_LE or S32_LE and eventually end up to FW internal errors. See bug like this for detail: https://github.com/thesofproject/sof/issues/3318

Comment (dbaluta) [reply tools/topology/sof/pipe-amp-ref-capture.m4:109]: @keyonjie care to reflect this explanation in the commit message?

Comment (keyonjie) [reply tools/topology/sof/pipe-amp-ref-capture.m4:109]: @dbaluta sure. Adding the pcm capability section but not the pcm_converter related one which actually is not changed in this PR.

**Review by dbaluta — APPROVED** [review]

File: tools/topology/m4/utils.m4 [hunk 2]
```diff
@@ -57,6 +57,15 @@ define(`COMP_FORMAT_VALUE',
 	$1, `s32le', `0x02',
 	)')
 
+dnl CAPABILITY_FORMAT_NAME(BUF_FMT)
+define(`CAPABILITY_FORMAT_NAME',
+`ifelse(
+	$1, `s16le', `S16_LE',
+	$1, `s24le', `S24_LE,S16_LE',
+	$1, `s32le', `S32_LE,S24_LE,S16_LE',
+	$1, `float', `FLOAT_LE',
```
Comment (lgirdwood) [inline tools/topology/m4/utils.m4:66]: What's the rational here, do you want to mask the format names so one name can be used to describe more than one format ? i.e. ```s32s24le``` would mean 32bit and 24 bit but NOT 16bit.

Comment (keyonjie) [reply tools/topology/m4/utils.m4:66]: No, I am not gonna change the naming here, instead, as written in the commit message, it adds helper to generate supported formats of the PCM capability from the pipeline format (buffer allocated format).

`S32_LE,S24_LE,S16_LE' is the format we used to denote the supported format of PCM capability in our m4.

So here, it means for:
pcm specified with s16_le, the buffers are 16bit (2Bytes) size allocated, the PCM can support S16_LE at run-time only.
pcm specified with s24_le, the buffers are 24bit size allocated, the PCM can support both S24_LE and S16_LE at run-time.
pcm specified with s32_le, the buffers are 32bit size allocated, the PCM can support all S32_LE, S24_LE and S16_LE at run-time.

File: tools/topology/m4/utils.m4 [hunk 3]
```diff
@@ -57,6 +57,15 @@ define(`COMP_FORMAT_VALUE',
 	$1, `s32le', `0x02',
 	)')
 
+dnl CAPABILITY_FORMAT_NAME(BUF_FMT)
```
Comment (lgirdwood) [inline tools/topology/m4/utils.m4:60]: ok, lets have the rules in the comment.

File: tools/topology/m4/utils.m4 (hunk 2 above)
Comment (lgirdwood) [reply tools/topology/m4/utils.m4:66]: can we rename the s32le names here, it's confusing with S32LE uppercase name as they are NOT the same thing.

**Review by lgirdwood — CHANGES_REQUESTED** [review]

File: tools/topology/m4/utils.m4 (hunk 2 above)
Comment (keyonjie) [reply tools/topology/m4/utils.m4:66]: > can we rename the s32le names here, it's confusing with S32LE uppercase name as they are NOT the same thing.

They are existed for component format, and the uppercase names are for the PCM format. If want to change, we need another PR for to clean them up as there are hundreds of place where using them.

File: tools/topology/m4/utils.m4 (hunk 3 above)
Comment (keyonjie) [reply tools/topology/m4/utils.m4:64]: Sure.

**Comment by keyonjie** [comment]
@lgirdwood updated.

**Comment by keyonjie** [comment]
> I checked with sof-apl-pcm512x.tplg and sof-cht-max98090.tplg. Looks S32_LE,S24_LE,S16_LE is assigned to wrong places. We need to carefully check MACRO again.
> 
> ```
> 	formats "S32_LE"
> 	rate_min "S24_LE"
> 	rate_max "S16_LE"
> 	channels_min "48000"
> 	channels_max "48000"
> 	periods_min "2"
> 	periods_max "2"
> 	period_size_min	"2"
> 	period_size_max	"4"
> 	buffer_size_min	"192"
> 	buffer_size_max	"16384"
> ```

Good catch, thanks for spotting this, updated now.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@keyonjie can you check CI JSL HDA test is failing.

**Comment by aiChaoSONG** [comment]
@lgirdwood Known Failures on ICL and GLK are not related to this PR. Failures on BSW_CYN will be fixed by https://github.com/thesofproject/sof/pull/3348

**Review by aiChaoSONG — APPROVED** [review]

**Pull request merged**